scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -201,108 +247,221 @@ def _highly_variable_genes_single_batch(
     A DataFrame that contains the columns
     `highly_variable`, `means`, `dispersions`, and `dispersions_norm`.
     """
-    X = adata.layers[layer] if layer is not None else adata.X
+    data = _get_obs_rep(adata, layer=layer)
     if flavor == "seurat":
-        X = X.copy()
+        data = data.copy()
         if "log1p" in adata.uns_keys() and adata.uns["log1p"].get("base") is not None:
-            X *= np.log(adata.uns["log1p"]["base"])
-        # use out if possible. only possible since we copy X
-        if isinstance(X, np.ndarray):
-            np.expm1(X, out=X)
+            data *= np.log(adata.uns["log1p"]["base"])
+        # use out if possible. only possible since we copy the data matrix
+        if isinstance(data, np.ndarray):
+            np.expm1(data, out=data)
         else:
-            X = np.expm1(X)
+            data = np.expm1(data)
 
-    mean, var = materialize_as_ndarray(_get_mean_var(X))
+    mean, var = _get_mean_var(data)
     # now actually compute the dispersion
     mean[mean == 0] = 1e-12  # set entries equal to zero to small value
     dispersion = var / mean
     if flavor == "seurat":  # logarithmized mean as in Seurat
         dispersion[dispersion == 0] = np.nan
         dispersion = np.log(dispersion)
         mean = np.log1p(mean)
+
     # all of the following quantities are "per-gene" here
-    df = pd.DataFrame()
-    df["means"] = mean
-    df["dispersions"] = dispersion
+    df = pd.DataFrame(
+        dict(zip(["means", "dispersions"], materialize_as_ndarray((mean, dispersion))))
+    )
+    df["mean_bin"] = _get_mean_bins(df["means"], flavor, n_bins)
+    disp_grouped = df.groupby("mean_bin", observed=True)["dispersions"]
```

Comment:
I think `df` should be passed into the `stats` functions and `disp_grouped` should be created within those methods.

It's a line of duplication, but I think it makes the code much easier to read, since all that code relies on a very particular relationship between these two values. Also the `disp_grouped` is directly derived from `df` so we can easily get to fewer arguments. 

Alternatively, you could write docstrings for the internal functions that explain the assumptions about how the arguments are related. But I think it's easier/ clearer/ shorter to just pass the dataframe.

Reply:
Or even better, just get rid of all the duplication by using a `_get_disp_stats` function that only receives the data frame